### ISSUE TITLE
CentOS 5.4 - yum module AttributeError "preconf"

### ISSUE THREAD

**Issue by eddgrant** [ansible/ansible#5700]
Raising as a result of [this discussion](https://groups.google.com/forum/#!topic/ansible-project/aBE51ToAC7A). I'm getting the following error when trying to execute the yum module to install an RPM on CentOS 5.4:

```
fatal: [dev.eddgrant.com] => failed to parse: SUDO-SUCCESS-owdhaivbcxarzwabxjtblacdvifhvhtk
Traceback (most recent call last):
  File "/home/ansible/.ansible/tmp/ansible-1390233348.21-255866138675253/yum", line 1766, in ?
    main()
  File "/home/ansible/.ansible/tmp/ansible-1390233348.21-255866138675253/yum", line 737, in main
    disablerepo, disable_gpg_check)
  File "/home/ansible/.ansible/tmp/ansible-1390233348.21-255866138675253/yum", line 659, in ensure
    my = yum_base(conf_file)
  File "/home/ansible/.ansible/tmp/ansible-1390233348.21-255866138675253/yum", line 123, in yum_base
    my.preconf.debuglevel=0
AttributeError: 'YumBase' object has no attribute 'preconf'
```

The task which causes this is as follows:

```
- name: Install Apache HTTPD Package
  yum: name=httpd state=latest
```

With the exception of the yum module all other tasks/ plays executed against this host complete successfully.

The system I'm running this from is: Ubuntu 12.10 (64 bit) running ansible 1.4.3
The system being provisioned is: CentOS 5.4 32 bit using JSON module: python-simplejson i386 Version: 2.0.9 Release: 1.el5.rf

Let me know if there's any other diagnostic information I can provide.

Cheers,

Edd

**Comment by jctanner** [comment]
@eddgrant what version of yum is on the centos host?

```
[root@el5 ~]# rpm -qi yum
Name        : yum                          Relocations: (not relocatable)
Version     : 3.2.22                            Vendor: CentOS
Release     : 40.el5.centos                 Build Date: Thu 26 Jul 2012 04:17:07 AM EDT
Install Date: Tue 08 Oct 2013 10:20:06 AM EDT      Build Host: builder10.centos.org
Group       : System Environment/Base       Source RPM: yum-3.2.22-40.el5.centos.src.rpm
Size        : 3429595                          License: GPLv2+
Signature   : DSA/SHA1, Thu 26 Jul 2012 04:22:27 AM EDT, Key ID a8a447dce8562897
URL         : http://yum.baseurl.org/
Summary     : RPM installer/updater
Description :
Yum is a utility that can check for and automatically download and
install updated RPM packages. Dependencies are obtained and downloaded
automatically prompting the user as necessary.
```

This patch would correct the traceback, but I'm not sure what else you would run into until I know the version of yum on the host:

```
diff --git a/library/packaging/yum b/library/packaging/yum
index f936b45..19e0e1c 100644
--- a/library/packaging/yum
+++ b/library/packaging/yum
@@ -133,8 +133,10 @@ def log(msg):
 def yum_base(conf_file=None, cachedir=False):

     my = yum.YumBase()
-    my.preconf.debuglevel=0
-    my.preconf.errorlevel=0
+    if hasattr(my.preconf, "debuglevel"):
+        my.preconf.debuglevel=0
+    if hasattr(my.preconf, "errorlevel"):
+        my.preconf.errorlevel=0
     if conf_file and os.path.exists(conf_file):
         my.preconf.fn = conf_file
     if cachedir or os.geteuid() != 0:
```

**Comment by eddgrant** [comment]
Hey, It's running 3.0.5:

```
    [egrant@centos5 ~]$ rpm -qi yum
    Name        : yum                          Relocations: (not relocatable)
    Version     : 3.0.5                             Vendor: eApps
    Release     : 5eapps                        Build Date: Mon 06 Oct 2008 06:32:27 PM EDT
    Install Date: Sat 10 Jul 2010 12:20:55 PM EDT      Build Host: centos-x86-build.development.eapps.com
    Group       : System Environment/Base       Source RPM: yum-3.0.5-5eapps.src.rpm
    Size        : 1380490                          License: GPL
    Signature   : (none)
    Packager    : Alves Adrian <aalves@eapps.com>
    URL         : http://linux.duke.edu/yum/
    Summary     : RPM installer/updater
    Description :
    Yum is a utility that can check for and automatically download and
    install updated RPM packages. Dependencies are obtained and downloaded
    automatically prompting the user as necessary.
```

**Comment by jimi-c** [comment]
@jctanner I think your proposed patch needs to check if the "my" variable has the "preconf" attribute.

<pre>
diff --git a/library/packaging/yum b/library/packaging/yum
index f936b45..577229c 100644
--- a/library/packaging/yum
+++ b/library/packaging/yum
@@ -133,10 +133,11 @@ def log(msg):
 def yum_base(conf_file=None, cachedir=False):
 
     my = yum.YumBase()
-    my.preconf.debuglevel=0
-    my.preconf.errorlevel=0
-    if conf_file and os.path.exists(conf_file):
-        my.preconf.fn = conf_file
+    if hasattr(my, 'preconf'):
+        my.preconf.debuglevel=0
+        my.preconf.errorlevel=0
+        if conf_file and os.path.exists(conf_file):
+            my.preconf.fn = conf_file
     if cachedir or os.geteuid() != 0:
         if hasattr(my, 'setCacheDir'):
             my.setCacheDir()
</pre>

@eddgrant can you let us know if either of these patches correct the issue for you?

**Comment by eddgrant** [comment]
Sure, happy to test, afraid I'll have to ask for guidance on how to build the patch (I'm an Ansible newbie and so am just using the pre-packaged debian packages in the Ubuntu repo at the moment).

If you can point me in the direction of some instructions I'll give it a go and will report back.

**Comment by jctanner** [comment]
@jimi-c oops, thanks ... just got a 5.1 system running to dig into this.

**Comment by jctanner** [comment]
reproduced the issue on 5.1

**Comment by jctanner** [comment]
@eddgrant it seems like you are using a third party yum package that eapps (hosting provider?) is maintaining? Yum 3.0.5 was last shipped with centos 5.1. Are you able to upgrade yum on the host(s) to 3.2.x ?

I can work past the initial traceback on 3.0.5, but then I run into a lot of other issues with the objects returned that would require some significant refactoring to the yum module.

**Comment by eddgrant** [comment]
Interesting, yeah eApps are the hosting provider of this VM. I'll see if I can re-jig the repo configs a little to bring in the default yum package. Will report back.

**Comment by eddgrant** [comment]
A bit of jiggery-pokery required to get back to the default CentOS repos but I've done it. I updated yum with `sudo yum install -y yum` which felt a bit like I was about to cause some sort of paradox, I'm still here though so I can only assume it was a success :-)

I'm now running yum 3.2.22 and can confirm that the problem is resolved by this upgrade. If, as you mentioned, yum 3.0.5 was last shipped with centos 5.1 then I can quite understand if you don't want to have to make such a significant re-factor simply to support such an ancient version. Will leave down to you as to whether you want to close this as a "won't fix" or whether to implement it. As far as I'm concerned however my problem is resolved.

Thanks for all help!

**Comment by jctanner** [comment]
@eddgrant yeah, let's close this. I'm glad you found a solution though.

**Comment by dbtucker** [comment]
Any other thoughts on a potential cause for this issue.   I'm seeing the problem on a new cloud provider's VM's (not Amazon or Google).   The image is CentOS 6.4 with yum version 3.2.29 ... but the symptom is exactly the same ("invalid output was: SUDO-SUCCESS-<long-random-string>"

What's weirder is that the problem only happens for packages from a non-CentOS repository.   I can log on to the system and run "sudo yum install -y <mypackage>", but the same operation via ansible fails.

**Comment by dbtucker** [comment]
Answered my own question with a few more searches.   This seems to be the issue reported at https://github.com/ansible/ansible/issues/5533

**Comment by 00willo** [comment]
I hit this yum module attribute error.   It was a "fresh" CentOS 5.2 install that I needed to replicate an existing system for some playbook testing.  So, yum was already at 3.2.22, and I still hit this exception.  I then updated the rpm-python package, which also updated rpm and 2 other deps.  After doing that, the playbook didn't halt on this exception.  Hope this helps someone else who might come across this.
